File: Numeros Pseudo-Aleatorios/Blum Blum Shub/BlumBlumShub.py
""" 
 PRNG .......: Blum-Blum-Shub
 Ano ........: 1986
 Autor ......: L. Blum, M. Blum & M. Shub
"""

import math
import time

def BlumBlumShub(X):
    X = (X * X) % 2467246241
    return int(X)

Xn1 = [0] * 10
Xn2 = [0] * 10
X1, X2 = 0, 0

print("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
print("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
print("PADRAO: Blum-Blum-Shub")
print("==========================================\n")
print("    Semente: 1234        Semente: TIME")
print("    -------------        -------------\n")

X1 = 1234
X2 = int(time.time())

for i in range(10):
    X1 = BlumBlumShub(X1)
    X2 = BlumBlumShub(X2)
    Xn1[i] = X1
    Xn2[i] = X2
    print(f"{Xn1[i]:17d}        {Xn2[i]:13d}")

File: Numeros Pseudo-Aleatorios/Blum Blum Shub/BlumBlumShub.c
/* 
 PRNG .......: Blum-Blum-Shub
 Ano ........: 1986
 Autor ......: L. Blum, M. Blum & M. Shub
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

unsigned int BlumBlumShub(unsigned int *X) {
  *X = (*X * *X) % 2467246241;
  return *X;
}

unsigned int Xn1[10];
unsigned int Xn2[10];
unsigned int X1, X2;

int main(void) {
	
  printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
  printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
  printf("PADRAO: Blum-Blum-Shub\n");
  printf("==========================================\n\n");
  printf("    Semente: 1234        Semente: TIME\n");
  printf("    -------------        -------------\n");

  X1 = 1234;
  X2 = (unsigned int)time(NULL);

  for (int i = 0; i <= 9; i++) {
    Xn1[i] = BlumBlumShub(&X1);
    Xn2[i] = BlumBlumShub(&X2);
    printf("%17u        %13u\n", Xn1[i], Xn2[i]);
  }

  return 0;
}
